src/workload.py

```diff
@@ -223,6 +224,31 @@ def _router_username(self) -> str:
         )
         return config["metadata_cache:bootstrap"]["user"]
 
+    def _wait_until_mysql_router_ready(self) -> None:
+        """Wait until a connection to MySQL Router is possible.
+
+        Retry every 5 seconds for up to 30 seconds.
+        """
+        logger.debug("Waiting until MySQL Router is ready")
+        self._charm.set_status(
```

Comment:
_you can safely ignore this_
Should, for the sake of user visibility, instead of using `set_status`  allow the handler set the status freely at the hook start and let `set_status/determine_status` at the hook's end?
My rational here is that, by using set_status at the start, we can potentially masquerade intermediate states due to the fixed priority in `_determine_status`, e.g. a maintenance triggered by tls setup will not be seen by the user case there's a blocked status in place.

Reply:
yes, so the questions boils to what's a better ui?
Other example*, imagine a long running action like backup/restore that will put the unit maintenance status - what it's best:
- to disallow the task when the unit is blocked?
- to hide the status change from the UI?
- to allow a lower priority status be set while the task runs?

imo, I'll go to the last one

*even though this example don't applies to this charm, we should strive for consistent UI across charm

Reply:
I am not in favor of "storing / keeping track" of a list of statuses. 
I agree with @paulomach's last point, in my opinion: when a hook gets invoked, the focus should be on the said hook, which means if an operation expected to report a status is expected, regardless of what it is => report it until the operation has completed, and only *upon completion*, reconcile statuses . i.e: 
```
def _on_start(event: StartEvent):
    set_unit_status(WaitingStatus("Starting mysql router..."))
    ....
    
    if self.unit.is_leader():
        reconcile_app_status()  # check the health of the **application**, and report new status (incl. ActiveStatus if all checks pass) which may be overriden by a subsequent hook call
        
    reconcile_unit_status()  # check the health of the **unit**, and report new status (incl. ActiveStatus if all checks pass) which may be overriden by a subsequent hook call
```

The "priority" of statuses should be the job of the charm developer to determine in the `reconcile_*_status` method / function.

Reply:
I am personally not in favor of "storing / keeping track" of a list of statuses. 
I agree with @paulomach's last point, in my opinion: when a hook gets invoked, the focus should be on the said hook, which means if an operation expected to report a status is invoked, regardless of what it is => report it until the operation has completed, and only *upon completion*, reconcile statuses . i.e: 
```
def _on_start(event: StartEvent):
    set_unit_status(WaitingStatus("Starting mysql router..."))
    ....
    
    if self.unit.is_leader():
        reconcile_app_status()  # check the health of the **application**, and report new status (incl. ActiveStatus if all checks pass) which may be overriden by a subsequent hook call
        
    reconcile_unit_status()  # check the health of the **unit**, and report new status (incl. ActiveStatus if all checks pass) which may be overriden by a subsequent hook call
```

The "priority" of statuses should be the job of the charm developer to determine in the `reconcile_*_status` method / function.

Reply:
> This requires the human operator to be aware of hook invocations
I'm not sure I understand what this means or the difference between the 2. If `on_install` is invoked as part of the lifecycle, the user should see `Installing .....`. Once the hook logic finishes, *recompute* the unit state and reflect it accordingly.
2 potential statuses, 1 on hook pre_execution logic, 1 after the hook logic completes.

Reply:
This is very interesting. I guess the charm can be in multiple modes (installing, operating, running an action), and you kind of want a different status to surface in each mode. Perhaps you could see "installing" and "running an action" both as a long-running-operation mode, and your status calculation would be something like this (untested pseudocode)

```python
def get_status(self):
    if charm.in_long_running_operation():
        # assume that long running operation will set status directly
        return ops.UnknownStatus()
    return this_components_status
```

That way if install was happening, it would set status directly with the sequence of status changes ("installing snap", blah, blah, blah), or if a long-running action was happening, same thing.

`in_long_running_operation` could be stored in stored state, or calculated, depending on the use case. Not entirely sure this is a good idea or if it would just add confusion.